WIP: Add runsc update command

kevinmingtarja opened the pull request:
This PR adds the `runsc update` command, which enables updating resource constraints of running containers. It is the runsc equivalent of `runc update`.

Closes #9554.

ayushr2 commented on runsc/cgroup/cgroup_v2.go:
```diff
@@ -233,6 +212,35 @@ func (c *cgroupV2) Join() (func(), error) {
+		found := false
+		for _, knownController := range c.Controllers {
+			if controllerName == knownController {
+				found = true
```
break here?

ayushr2 commented on runsc/cmd/update.go:
```diff
@@ -0,0 +1,215 @@
+
+// SetFlags implements subcommands.Command.SetFlags.
+func (u *Update) SetFlags(f *flag.FlagSet) {
+	f.StringVar(&u.resources, "resources", "", `path to the file containing the resources to update or '-' to read from the standard input
```
Note that the default value means that resources will be read from the remaining flags.

ayushr2 reviewed (commented):
Looks good so far!

ayushr2 commented:
@kevinmingtarja Do you have a chance to revive this?

kevinmingtarja commented:
> @kevinmingtarja Do you have a chance to revive this?

ah yes sorry, i just finished my finals and i can start working on this again!

kevinmingtarja commented:
remove stale

kevinmingtarja commented:
remove stale

kevinmingtarja commented:
Hi @ayushr2, sorry for the super long delay! This PR is ready for review now.

To recap:
- the main additions are `container.Set()` and `Cgroup.Set()` (for all its types, `cgroupv1`, `cgroupv2`, `cgroupSystemd`)
- added unit tests at `cgroup_v2_test.go` and `systemd_test.go`
- modified existing `cgroupv1` and `cgroupv2` E2E tests at `cgroup_test.go`
- tested manually as well on a VM with cgroupv2 and systemd, both via Docker and `runsc` directly

ayushr2 commented on runsc/cgroup/cgroup.go:
```diff
@@ -316,6 +316,7 @@ func loadPathsHelper(cgroup, mountinfo io.Reader, unified bool) (map[string]stri
 // Cgroup represents a cgroup configuration.
 type Cgroup interface {
 	Install(res *specs.LinuxResources) error
+	Set(res *specs.LinuxResources) error
```
IMO this should be `Update()`, to make it clear that the cgroup limits are being updated after initialization. This is necessary to differentiate it from `Install()`, which also "sets" the limits upon initialization.

kevinmingtarja replied:
makes sense, updated.

ayushr2 commented on runsc/cgroup/cgroup.go:
```diff
@@ -542,6 +543,35 @@ func (c *cgroupV1) Install(res *specs.LinuxResources) error {
+				}
+				log.Infof("Skipping cgroup %q, path doesn't exist", key)
+				continue
+			} else if err != nil {
```
`if err != nil` is always true here since it was checked on line 555.

I think the `else` clause is not needed at all since the previous if condition ends with `continue`. You can just return error here (helps save indentation).

kevinmingtarja replied:
ah right good catch, i agree.

ayushr2 commented on runsc/cgroup/cgroup_v2.go:
```diff
@@ -140,33 +140,43 @@ func (c *cgroupV2) Install(res *specs.LinuxResources) error {
+
+// Set sets the cgroup resources.
+func (c *cgroupV2) Set(res *specs.LinuxResources) error {
+	log.Debugf("Setting cgroup resources for %q", c.MakePath(""))
```
cache the value of `c.MakePath("")` so it can be re-used on line 167.

kevinmingtarja replied:
updated.

ayushr2 commented on runsc/cgroup/systemd.go:
```diff
@@ -135,6 +142,25 @@ func (c *cgroupSystemd) Install(res *specs.LinuxResources) error {
 			return fmt.Errorf("mandatory cgroup controller %q is missing for %q", controllerName, c.Path)
 		}
 	}
+
+	ctx := context.Background()
+	conn, err := systemdDbus.NewWithContext(ctx)
+	if err != nil {
+		return err
+	}
+	c.dbusConn = conn
+	// Our systemd units are transient, hence runtime=true.
+	if err := c.dbusConn.SetUnitPropertiesContext(ctx, c.unitName(), true /* runtime */, c.properties...); err != nil {
+		var derr dbus.Error
+		if errors.As(err, &derr) {
+			// We have not called Join() yet, so we cannot set the properties as the unit does not exist.
+			if strings.Contains(derr.Name, "org.freedesktop.systemd1.NoSuchUnit") {
+				return nil
+			}
+		}
+		return fmt.Errorf("error setting systemd unit properties: %v", err)
+	}
+
```
These changes are new. Should these also be execute for `Install()`? Or only for the `runsc update` case? If the former, I think this should be contributed as a separate PR as it changes behavior outside of `runsc update`.

kevinmingtarja replied:
good qn, it should only be executed for the `runsc update` case actually.

I had mistakenly called it during `Install()`, but @manninglucas caught that mistake. so i'll fix that.

ayushr2 reviewed (commented):
Thanks for the work!

Please squash your commits into 1, otherwise all the intermediate commits will also be merged into master once this is accepted.

manninglucas commented on runsc/cgroup/systemd.go:
```diff
@@ -111,12 +111,19 @@ func (c *cgroupSystemd) Install(res *specs.LinuxResources) error {
 	// For compatibility with runc.
 	c.addProp("DefaultDependencies", false)
 
+	return c.Set(res)
```
At this point we haven't setup the unit yet. That happens during `Join`. You can just return nil here.

kevinmingtarja replied:
got it, i should probably remove this block too then from the `Update()` method below:
```
var derr dbus.Error
if errors.As(err, &derr) {
	// We have not called Join() yet, so we cannot set the properties as the unit does not exist.
	if strings.Contains(derr.Name, "org.freedesktop.systemd1.NoSuchUnit") {
		return nil
	}
}
```

ayushr2 commented on runsc/cgroup/cgroup_v2.go:
```diff
@@ -170,6 +171,38 @@ func (c *cgroupV2) Install(res *specs.LinuxResources) error {
 }
 
+// Update updates the cgroup resources.
+func (c *cgroupV2) Update(res *specs.LinuxResources) error {
```
line 142-167 can be replaced with just call to `c.Update(res)`. The code is identical.

ayushr2 commented on runsc/cgroup/systemd.go:
```diff
@@ -138,6 +139,48 @@ func (c *cgroupSystemd) Install(res *specs.LinuxResources) error {
+// Update updates the cgroup resources of an existing systemd unit.
+func (c *cgroupSystemd) Update(res *specs.LinuxResources) error {
+	log.Debugf("Updating systemd cgroup resource controller under %v", c.Parent)
+	for controllerName, ctrlr := range controllers2 {
+		// First check if our controller is found in the system.
+		found := false
+		for _, knownController := range c.Controllers {
+			if controllerName == knownController {
+				found = true
+				break
+			}
+		}
+		if found {
+			props, err := ctrlr.generateProperties(res)
+			if err != nil {
+				return err
+			}
+			c.properties = append(c.properties, props...)
+			continue
+		}
+		if ctrlr.optional() {
+			if err := ctrlr.skip(res); err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("mandatory cgroup controller %q is missing for %q", controllerName, c.Path)
+		}
+	}
```
Refactor this into a new function, maybe `update()` and replace line 114-138 with call to `c.update(res)`.

kevinmingtarja replied:
sounds good, i think let's name it `updateControllerProps()`, to make it similar to `addProp()` and `addIOProps()` which already exists in `cgroupSystemd`.

ayushr2 commented on runsc/container/container.go:
```diff
@@ -552,6 +552,39 @@ func Run(conf *config.Config, args Args) (unix.WaitStatus, error) {
 }
 
+// Set sets the resources of a running container as configured.
+func (c *Container) Set(res *specs.LinuxResources) error {
```
Change `Set` -> `Update`?

ayushr2 reviewed (commented):
Thanks mostly looks good!

ayushr2 commented on runsc/cmd/update.go:
```diff
@@ -0,0 +1,296 @@
+			strval string
+			dest   **int64
+		}{
+			{u.memory, &r.Memory.Limit},
+			{u.memoryReservation, &r.Memory.Reservation},
+			{u.memorySwap, &r.Memory.Swap},
```
This loop is identical to the loop above. Just add these entries above and get rid of this loop.

kevinmingtarja replied:
good catch, fixed!

ayushr2 commented on runsc/cmd/update.go:
```diff
@@ -0,0 +1,281 @@
+		}
+	}
+
+	if prev.BlockIO != nil {
```
I have merged this change. Since this is what runc does, I assume this is the right thing to do.
